### ISSUE TITLE
Creating "multi-channel" MNIST digits

### ISSUE THREAD

**Issue by bamford** [aidenrolfe/ARG#11]
Your input data is `x`, and array of shape `(n, w, h, 1)`, i.e. `n` images of size `(w, h)` with a single "channel".

To make this multi-channel in the simplest way, we want an array that has the same content, but duplicated `c` times along the final axis, i.e. of shape `(n, w, h, c)`, where `x[:, :, :, 0]` equals `x[:, :, :, 1]`, etc. 

To do this naively, one could create a new array of the size we want and copy the data over:
```python
x_new = np.zeros((n, w, h, c))
for i in range(c):
    x_new[..., i] = x
```

However, a more elegant solution is to make use of array “broadcasting”, see [here](https://numpy.org/doc/stable/user/basics.broadcasting.html). When two arrays of different shapes are operated on (added, multiplied, ...), one is "broadcasted" over the other (if the shapes are compatible). Getting used to the broadcasting rules takes some time and I usually try out a little test to make sure. Anyway, this would have the desired effect:
```python
x_new = x * np.ones(c)
```
For c=3, you can think of this as being a normal colour image of each digit. However, as the red, green and blue channels are identical, the images would look grey. 

Let's mix things up a bit. Say you want the digits all different "colours". We first need an array giving the r, g, b level for each of the `n` images:
```python
colours = np.random.uniform(size=(n, c))
```
This doesn't have a shape that is compatible with `x`, so we need to add in the missing width and height axes, like this:
```python
colours = colours[:, None, None, :]
```
which turns it into an array of shape `(n, 1, 1, c)`.

Now we can multiply them:
```python
x_new = x * colours
```
producing the `(n, w, h, c)` array we are after.

This is what Aiden has been doing to produce the galaxy data. If you give matplotlib's `imshow` a 3 channel image it will produce a colour image, so you can do this and test it works.

Now, we come to the issue of noise. If we just want to add noise (such that it is different for each channel), then you can just add it in exactly the same way as before:
```python
x_noisy = x_new + noise_factor * np.random.normal(size=x_new.shape)
```

It isn't necessarily desirable for the tests you are doing, but if you want to get fancy and have some channels noisier than others, then you could do something like:
```python
noise_levels = np.random.uniform(c)
noise = np.random.normal(size=x_new.shape)
noise = noise_levels * noise
x_noisy = x_new + noise
```
In the third line we have used broadcasting again, but as the shapes of `x_new` and `noise_levels` both end with `c`, the result is the same shape as `x_new`.

**Comment by larissapalethorpe** [comment]
I implemented the code above into the VAE.py file which worked well I just had to use `n, w, h, c = x_train.shape` to get the appropriate values x_train, and the repeated the process for x_test. I also had to changed encoder_inputs to `shape=(28,28,3)`, and running for 10 epochs gave 
<img width="1153" alt="Screenshot 2021-03-15 at 14 35 07" src="https://user-images.githubusercontent.com/78367556/111170678-1a71f380-859c-11eb-9339-f8e94a209e3d.png">

The reconstructions are gray, however which I think is because I haven't edited the y_test or y_train to be colourful? I was just wondering if this is what needs to be done? I've edited the VAE.py file so you can see what I'm currently doing

**Comment by bamford** [comment]
Looking good, just a few things...

### Number of output channels

The output from the network still contains only one channel. You need to change the number of output "features" in [this line](https://github.com/aidenrolfe/ARG/blob/98ab123e290fefae48a5766ebe1e9d69cd6b5f6f/VAE.py#L147) to `c` (3 in your current code):
```python
decoder_outputs = layers.Conv2DTranspose(c, 3, activation="sigmoid", padding="same")(x)
```
By the way, note that `c`, `w` and `h` must be the same for all the data, so you don't need separate variables for train and test.
You probably also want to use `c`, rather than `3` in [this line](https://github.com/aidenrolfe/ARG/blob/98ab123e290fefae48a5766ebe1e9d69cd6b5f6f/VAE.py#L100), so that you can change `c` easily.

### Colour learning happens late in training 

I tried it out, and found that the network learns grayscale versions of the numbers first, and only later in the training learns the colours. So, only training for a small number of epochs results in gray images. Note that to encode the colours takes 3 values, so three dimensions in the latent space, leaving only two for shape variations, so there will be a trade-off in performance (although maybe not so bad when restricting to one digit class). Increasing the number of latent dimensions will improve performance.

### Noise

You can see that your current input images are very noisy. This will make it more challenging for the network to learn. Might be best to turn down the noise factor a bit for now.

### Condition for one digit

Your current code is restricting to one digit, but still providing a condition. The condition label will always be the same, so if you are restricting to one digit, you don't need to use the condition. 

### Small training set

Given the amount of variation added in by the colours, the training set is now on the small side, especially when limiting to just one digit (which reduces the dataset by a factor of ~10). This may also make it harder for the network to learn.

### Model saving

It's a good to start [saving your models](https://www.tensorflow.org/tutorials/keras/save_and_load#save_the_entire_model) (`vae.save(filename)`) after fitting. Then you can reload them and remake plots without having to retrain every time.

**Comment by bamford** [comment]
I've been playing... I've taken your code and refactored it into functions – hopefully you can still see what's what – and run it on Google Colab here: https://colab.research.google.com/drive/1QAxMIhD2c8Szf38B7c03NL6oboQke3eB?usp=sharing

**Comment by larissapalethorpe** [comment]
I implemented the changes listed above (although haven't had a chance to run the colab version listed yet). I ran for 8 latent dimensions, beta of 0.1, for a maximum of 200 epochs but with the EarlyStopping callback. Training stopped over 69 epochs (`loss: 7.8759 - reconstruction_loss: 6.5995 - kl_loss: 1.2780 - val_loss: 8.4796 - val_reconstruction_loss: 7.2034 - val_kl_loss: 1.2912`). I also took out the digit filter but set `conditional='yes'`. The reconstructions are looking pretty good though! Some of the colours are quite off though which is interesting
<img width="1146" alt="Screenshot 2021-03-16 at 12 08 04" src="https://user-images.githubusercontent.com/78367556/111307068-e1e02180-8650-11eb-96f1-bf30c476ce98.png">
